autogluon/utils/tabular/ml/trainer/abstract_trainer.py

```diff
@@ -867,225 +574,636 @@ def refit_single_full(self, X=None, y=None, X_val=None, y_val=None, models=None)
+            self.save_model(model=model)
+        except TimeLimitExceeded:
+            logger.log(20, f'\tTime limit exceeded... Skipping {model.name}.')
+            # logger.log(20, '\tTime wasted: ' + str(time.time() - fit_start_time))
+            del model
+        except NotEnoughMemoryError:
+            logger.warning(f'\tNot enough memory to train {model.name}... Skipping this model.')
+            del model
+        except NoValidFeatures:
+            logger.warning(f'\tNo valid features to train {model.name}... Skipping this model.')
+            del model
+        except Exception as err:
+            if self.verbosity >= 1:
+                traceback.print_tb(err.__traceback__)
+            logger.exception('Warning: Exception caused %s to fail during training... Skipping this model.' % model.name)
+            logger.log(20, err)
+            del model
+        else:
+            self._add_model(model=model, stack_name=stack_name, level=level)
+            model_names_trained.append(model.name)
+            if self.low_memory:
+                del model
+        return model_names_trained
+
+    def _add_model(self, model: AbstractModel, stack_name: str = 'core', level: int = 0):
+        """
+        Registers the fit model in the Trainer object. Stores information such as model performance, save path, model type, and more.
+        To use a model in Trainer, self._add_model must be called.
+
+        If self.low_memory, then the model object will be deleted after this call. Use Trainer directly to leverage the model further.
+
+        Parameters
+        ----------
+        model : AbstractModel
+            Model which has been fit. This model will be registered to the Trainer.
+        stack_name : str, default 'core'
+            Stack name to assign the model to. This is used for advanced functionality.
+        level : int, default 0
+            Stack level of the stack name to assign the model to. This is used for advanced functionality.
+            The model's name is appended to self.models_level[stack_name][level]
+            The model's base_models (if it has any) must all be a lower level than the model.
+        """
+        if model.val_score is not None and stack_name != self.distill_stackname:  # TODO: may want to avoid hard-coding logic into specific stack names
+            logger.log(20, '\t' + str(round(model.val_score, 4)) + '\t = Validation ' + self.eval_metric.name + ' score')
+        if model.fit_time is not None:
+            logger.log(20, '\t' + str(round(model.fit_time, 2)) + 's' + '\t = Training runtime')
+        if model.predict_time is not None:
+            logger.log(20, '\t' + str(round(model.predict_time, 2)) + 's' + '\t = Validation runtime')
+        # TODO: Add to HPO
+        # TODO/FIXME: what if it isn't valid?
+        if model.is_valid():
```

Comment:
should there be warning if not?

Reply:
removed this if statement, it isn't necessary